pallets/pallet-bonded-coins/src/lib.rs

```diff
@@ -0,0 +1,1577 @@
+					freeze_error.into()
+				})
+				.map_err(|_| TokenError::FundsUnavailable)?;
+
+			// With amount = max_value(), this trait implementation burns the reducible
+			// balance on the account and returns the actual amount burnt
+			let burnt: U256 = T::Fungibles::burn_from(
+				asset_id.clone(),
+				&who,
+				Bounded::max_value(),
+				WithdrawalPrecision::BestEffort,
+				Fortitude::Force,
+			)?
+			.into();
+
+			ensure!(!burnt.is_zero(), TokenError::FundsUnavailable);
+
+			let sum_of_issuances = pool_details
+				.bonded_currencies
+				.into_iter()
+				.fold(U256::from(0u8), |sum, id| {
+					sum.saturating_add(T::Fungibles::total_issuance(id).into())
+				})
+				// Add the burnt amount back to the sum of total supplies
+				.checked_add(burnt)
+				.ok_or(ArithmeticError::Overflow)?;
+
+			defensive_assert!(
+				sum_of_issuances >= burnt,
+				"burnt amount exceeds the total supply of all bonded currencies"
+			);
+
+			let amount: CollateralCurrenciesBalanceOf<T> = burnt
+				.checked_mul(total_collateral_issuance.into())
+				// As long as the balance type is half the size of a U256, this won't overflow.
+				.ok_or(ArithmeticError::Overflow)?
+				.checked_div(sum_of_issuances)
+				// Because sum_of_issuances >= burnt > 0, this is theoretically impossible
+				.ok_or_else(|| {
+					log::error!(
+						target: LOG_TARGET,
+						"Sum of issuance is zero. Pool_id: {:?}, account: {:?} Burnt: {:?}, Total Collateral Issuance: {:?}, Sum of Issuances: {:?}", 
+						pool_id,
+						who,
+						burnt,
+						total_collateral_issuance,
+						sum_of_issuances
+					);
+					Error::<T>::Internal
+				})?
+				.checked_into()
+				// Also theoretically impossible, as the result must be <= total_collateral_issuance
+				// if burnt <= sum_of_issuances, which should always hold true
+				.ok_or_else(|| {
+					log::error!(
+						target: LOG_TARGET,
+						"Could not cast U256 into CollateralCurrency. Pool_id: {:?}, account: {:?} Burnt: {:?}, Total Collateral Issuance: {:?}, Sum of Issuances: {:?}", 
+						pool_id,
+						who,
+						burnt,
+						total_collateral_issuance,
+						sum_of_issuances
+					);
+					Error::<T>::Internal
+				})?;
+
+			if amount.is_zero()
+				|| T::CollateralCurrencies::can_deposit(
+					pool_details.collateral_id.clone(),
+					&who,
+					amount,
+					Provenance::Extant,
+				)
+				.into_result()
+				.is_err()
+			{
+				// Funds are burnt but the collateral received is not sufficient to be deposited
+				// to the account. This is tolerated as otherwise we could have edge cases where
+				// it's impossible to refund at least some accounts.
+				return Ok(Some(T::WeightInfo::refund_account(currency_count.to_owned())).into());
+			}
+
+			let transferred = T::CollateralCurrencies::transfer(
+				pool_details.collateral_id,
+				&pool_account,
+				&who,
+				amount,
+				Preservation::Expendable,
+			)?; // TODO: check edge cases around existential deposit
```

Comment:
```suggestion
			)?;
```
I think we've done this and handle issues around EDs with the early return just above

Reply:
f101524168b6f1b0e1d698d100124c9a3fe0d0f0